compiler/rustc_lint/src/mutable_transmutes.rs

```diff
@@ -0,0 +1,718 @@
+    }
+}
+
+#[derive(PartialEq)]
+enum CollectFields {
+    Yes,
+    No,
+}
+
+fn collect_fields<'tcx>(
+    cx: &LateContext<'tcx>,
+    ty: Ty<'tcx>,
+    offset: Size,
+    callback: &mut impl FnMut(Size, Ty<'tcx>, &FieldsBreadcrumbs) -> CollectFields,
+    fields_breadcrumbs: &mut FieldsBreadcrumbs,
+) {
+    if callback(offset, ty, &fields_breadcrumbs) == CollectFields::No {
+        return;
+    }
+
+    match ty.kind() {
+        &ty::Adt(adt_def, substs) if adt_def.is_struct() => {
+            let Ok(layout) = cx.layout_of(ty) else {
+                return;
+            };
+            debug!(?offset, ?layout, "collect_fields struct");
+            for (field_idx, field_offset, field_ty) in
+                struct_fields(cx, layout.layout, adt_def, substs)
+            {
+                fields_breadcrumbs.push_with(field_idx, |fields_breadcrumbs| {
+                    collect_fields(
+                        cx,
+                        field_ty,
+                        offset + field_offset,
+                        callback,
+                        fields_breadcrumbs,
+                    )
+                });
+            }
+        }
+
+        &ty::Tuple(tys) => {
+            let Ok(layout) = cx.layout_of(ty) else {
+                return;
+            };
+            debug!(?offset, ?layout, "collect_fields tuple");
+            for (field_idx, field_offset, field_ty) in tuple_fields(layout.layout, tys) {
+                fields_breadcrumbs.push_with(field_idx, |fields_breadcrumbs| {
+                    collect_fields(
+                        cx,
+                        field_ty,
+                        offset + field_offset,
+                        callback,
+                        fields_breadcrumbs,
+                    )
+                });
+            }
+        }
+
+        &ty::Array(ty, len)
+            if let Some(len) = len.try_to_target_usize(cx.tcx)
+                && len != 0 =>
+        {
+            debug!(?offset, "collect_fields array");
+            fields_breadcrumbs.push_with(FieldIdx::from_u32(0), |fields_breadcrumbs| {
+                collect_fields(cx, ty, offset, callback, fields_breadcrumbs)
+            });
+        }
+
+        _ => {}
+    }
+}
+
+trait Field {
+    fn start_offset(&self) -> Size;
+    fn end_offset(&self) -> Size;
+}
+
+#[derive(Debug)]
+enum CheckOverlapping {
+    Yes,
+    No,
+}
+
+/// `check_overlap_with` must be sorted by starting offset.
+fn check_overlapping<'tcx>(
+    cx: &LateContext<'tcx>,
+    check_overlap_with: &[impl Field],
+    ty: Ty<'tcx>,
+    offset: Size,
+    on_overlapping: &mut impl FnMut(Ty<'tcx>, usize, &FieldsBreadcrumbs) -> CheckOverlapping,
+    fields_breadcrumbs: &mut FieldsBreadcrumbs,
+) {
+    let Ok(layout) = cx.layout_of(ty) else {
+        return;
+    };
+    debug!(?offset, ?layout, "check_overlapping");
+
+    // Then, for any entry that we overlap with (there can be many as our size can be bigger than one,
```

Comment:
```suggestion
    // Then, for any entry that we overlap with (there can be many because our size can be bigger than one,
```
(I had to read this sentence twice to understand it. This is a pet peeve of mine, "because" is almost always clearer than "as".)